File: iOS8_Adv_Training/CollectionView/CollectionView/ViewController.m
//
//  ViewController.m
//  CollectionView
//
//  Created by 新居雅行 on 2014/10/29.
//  Copyright (c) 2014年 msyk.net. All rights reserved.
//

#import "ViewController.h"
#import "DataCellView.h"

#define CELL_WIDTH 42.0
#define CELLS_PAR_LINE 7
#define END_OF_MONTH @[@0,@31,@28,@31,@30,@31,@30,@31,@31,@30,@31,@30,@31]

@interface ViewController ()
@property (nonatomic) NSInteger startDateBias;
@property (nonatomic) NSInteger endDayNum;
@property (nonatomic) NSInteger lines;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    self.year = 2013;
    self.month = 10;
    
    self.navigationItem.title
    = [NSString stringWithFormat: @"%d/%d", self.year, self.month];
    
    UICollectionViewFlowLayout *fl = (UICollectionViewFlowLayout *)self.collectionViewLayout;
    fl.minimumInteritemSpacing = (320.0 - CELL_WIDTH * CELLS_PAR_LINE) / (CELLS_PAR_LINE - 1);
    fl.sectionInset = UIEdgeInsetsMake(1.0, 0, 1.0, 0);
    
    NSDateFormatter *dtFormatter = [[NSDateFormatter alloc]init];
    [dtFormatter setDateFormat: @"yyyy-MM-dd"];
    NSDate *firstDate = [dtFormatter dateFromString:
                         [NSString stringWithFormat: @"%4d-%2d-01", self.year, self.month]];
    NSCalendar *cal = [NSCalendar currentCalendar];
    NSDateComponents *comps = [cal components: NSCalendarUnitWeekday
                                     fromDate: firstDate];
    self.startDateBias = comps.weekday - 1;
    self.endDayNum = [END_OF_MONTH[self.month] integerValue];
    self.lines = (self.endDayNum + self.startDateBias) / CELLS_PAR_LINE + 1;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - UICollectionViewDataSource

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView
                  cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
#ifdef DEBUG
    NSLog( @"%s", __FUNCTION__ );
#endif
    DataCellView *cell
    = [collectionView dequeueReusableCellWithReuseIdentifier: @"DATECELL"
                                                forIndexPath: indexPath];
    //cell.dateLabel.text = [NSString stringWithFormat: @"%d", indexPath.item];
    NSInteger dayNum = indexPath.item
    + indexPath.section * CELLS_PAR_LINE - self.startDateBias + 1;
    if (dayNum > 0 && dayNum <= self.endDayNum) {
        cell.dateLabel.text = [NSString stringWithFormat: @"%d", dayNum];
    } else {
        cell.dateLabel.text = @".";
    }
    return cell;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView
     numberOfItemsInSection:(NSInteger)section
{
#ifdef DEBUG
    NSLog( @"%s", __FUNCTION__ );
#endif
    return CELLS_PAR_LINE;
}
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
#ifdef DEBUG
    NSLog( @"%s", __FUNCTION__ );
#endif
    return self.lines;
}
@end

File: iOS8_Adv_Training/CollectionView_Swift/CollectionView_Swift/ViewController.swift
//
//  ViewController.swift
//  CollectionView_Swift
//
//  Created by 新居雅行 on 2015/01/29.
//  Copyright (c) 2015年 msyk.net. All rights reserved.
//

import UIKit

class ViewController: UICollectionViewController,
UICollectionViewDataSource, UICollectionViewDelegate {
    
    var year: Int = 2015
    var month: Int = 1
    
    private var startDateBias: Int = 0
    private var endDayNum: Int = 0
    private var lines: Int = 0
    
    private let end_of_month: [Int]
    = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    private let cells_per_line: Int = 7
    private let cell_width: Float = 42.0
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = "\(year) \(month)"
        
        let fl = collectionViewLayout as! UICollectionViewFlowLayout
        let space = (320.0 - cell_width * Float(cells_per_line))
            / Float(cells_per_line - 1)
        fl.minimumInteritemSpacing = CGFloat(space)
        fl.sectionInset = UIEdgeInsets(top: 1.0, left: 0.0, bottom: 1.0, right: 0.0)
        
        let dtFormatter = NSDateFormatter()
        dtFormatter.dateFormat = "yyyy-MM-dd"
        let firstDate = dtFormatter.dateFromString(
            NSString(format:"%4d-%2d-01", year, month) as String)
        let cal = NSCalendar.currentCalendar()
        let comps = cal.components(
            NSCalendarUnit.WeekdayCalendarUnit,
            fromDate: firstDate!)
        startDateBias = comps.weekday - 1
        endDayNum = end_of_month[month]
        lines = (endDayNum + startDateBias) / cells_per_line
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - UICollectionViewDataSource
    override func collectionView(
        collectionView: UICollectionView,
        cellForItemAtIndexPath indexPath: NSIndexPath)
        -> UICollectionViewCell {
            let cell = collectionView.dequeueReusableCellWithReuseIdentifier(
                "DATACELL", forIndexPath: indexPath) as! DataCellView
            let dayNum = indexPath.item
                + indexPath.section * cells_per_line - startDateBias + 1
            if dayNum > 0 && dayNum <= endDayNum {
                cell.dataLabel?.text = "\(dayNum)"
            } else {
                cell.dataLabel?.text = "."
            }
            return cell
    }
    
    override func collectionView(
        collectionView: UICollectionView,
        numberOfItemsInSection section: Int) -> Int {
            return cells_per_line
    }
    
    override func numberOfSectionsInCollectionView(
        collectionView: UICollectionView) -> Int {
            return lines
    }
    
}

